src/Schema/Table.php

```diff
@@ -502,11 +503,24 @@ public function addForeignKeyConstraint(
         array $options = [],
         ?string $name = null,
     ): self {
-        $name ??= $this->_generateIdentifierName(
-            array_merge([$this->getName()], $localColumnNames),
-            'fk',
-            $this->_getMaxIdentifierLength(),
-        );
+        if ($name === null) {
+            $name = $this->_generateIdentifierName(
+                array_merge([$this->getName()], $localColumnNames),
+                'fk',
+                $this->_getMaxIdentifierLength(),
+            );
+            if ($this->needsAlternativeForeignKeyName($this->normalizeIdentifier($name), $foreignTableName)) {
```

Comment:
Looking at the complexity of the solution, I don't think the DBAL should try to accommodate any user input at any cost. The case described in the issue is very theoretical case, while the complexity being added is practical.

Instead, we might detect a collision and emit a deprecation notice (something like "Overwriting an existing foreign key constraint is deprecated"). In 5.0.x, until fixed properly, this will result in an exception.